### ISSUE TITLE
Saving a model in multiprocessing

### ISSUE THREAD

**Issue by ShaniGam** [ikostrikov/pytorch-a2c-ppo-acktr#106]
When using multiprocessing in A2C, each environment has its own interaction and therefore its reward. Let's say I have an environment in which I get 7000 points and in other 0, how can I restore the 7000 points environment during testing when loading the model?

**Comment by timmeinhardt** [comment]
Maybe my answer in #105 also helps you to understand your confusion here. First of all, we do not save models of the different environments but the agent model. In the case of A2C it is one single model for all environments which is constantly updated. I guess there are different possibilities for testing. You could either load the last policy model or the one which achieved the largest rewards. Although for the latter you should probably not consider the raw reward but the normalized, running reward. Since we are working with a stochastic policy a high raw reward doesn't necessarily mean that the policy model already to reproduce that behaviour.

**Comment by ShaniGam** [comment]
First, thanks for replying to my questions.
Second, right now I'm using the models with highest mean rewards. Unfortunately, I never get the max results I see on training. When I test a trained model I always get the same result (when using the same model) even though I use the sample function. I guess a good solution would be testing it during training with one agent after all the agents are done updating and saving it afterward.

**Comment by timmeinhardt** [comment]
From my understanding, it is definitely possible to see higher raw rewards during training compared to the saved model during testing. Probably your model did not yet converge to produce these high rewards reliably and the rewards you saw were just due to the stochasticity during training. 
It depends on your particular problem if you want to use a stochastic agent during testing, but you should be able to deploy a stochastic model with `deterministic=False` as argument for the `act` method.
I am not sure about your last sentence because there is only one agent that is getting updated on a batch of environments. What I do during training, is monitor both the deterministic and non-deterministic interaction of the current agent model with the environment.

**Comment by timmeinhardt** [comment]
There are many possible (training related) reasons for why your agent behaves in such a way. It seems there is something not yet working with your training, e.g, hyperparameters etc. If there is such a huge gap between a deterministic and non-deterministic agent you could also check how a random (not trained, non-deterministic) agent performs with respect to the reward. This gives you a good baseline for measuring the training success. A more detailed analysis of your problem probably goes beyond the scope of this issue.
